### PR TITLE
Move EffectCache to GraphicsDevice

### LINKED ISSUE #3449: Request: remove effects from Effect.EffectCache when effect is disposed
Effects are cached in a private static field in `Effect.EffectCache`. But the only code that calls the internal `Effect.FlushCache()` is `Game.Dispose()`.

It took me a while to figure out that this is the root cause of why some of my visual tests are failing. I'm not using a `Game` object - I'm just creating lower-level objects like `GraphicsDevice` directly. But because I'm disposing the `GraphicsDevice` and other graphics resources after each test, any effects left in the cache are unusable. Any subsequent tests that use the same `Effect` don't work.

I'm not sure what the best solution is - it seems like cached effects should be removed when the `GraphicsDevice` used to create them (rather than the `Game`) is disposed? Or at least, when the `Effect` itself is disposed (I think `GraphicsDevice.Dispose` already calls `Dispose` on `GraphicsResource` objects, so that may be enough).

For now, I've worked around it with this, but it's not pretty:

``` csharp
typeof(Effect).GetMethod("FlushCache", BindingFlags.NonPublic | BindingFlags.Static).Invoke(null, null);
```

**EDIT**: I've just run into a similar issue with state objects - I get exceptions because they are associated with a different graphics device. Basically, I think I need to be able to [what Game is doing here](https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Game.cs#L120) (or even better, have it happen implicitly when `GraphicsDevice` is disposed):

``` csharp
if (_graphicsDeviceManager != null)
{
    Effect.FlushCache();
    BlendState.ResetStates();
    DepthStencilState.ResetStates();
    RasterizerState.ResetStates();
    SamplerState.ResetStates();

    ...
}
```

### PR DESCRIPTION
**Pull request by tgjones**
Fixes #3449.

### CODE REVIEW

File: MonoGame.Framework/Graphics/GraphicsDevice.cs [hunk 1]
```diff
@@ -413,6 +421,9 @@ protected virtual void Dispose(bool disposing)
                     // Dispose of all remaining graphics resources before disposing of the graphics device
                     GraphicsResource.DisposeAll();
 
+                    // Clear the effect cache.
+                    EffectCache.Clear();
```
Comment (tomspilman) [inline MonoGame.Framework/Graphics/GraphicsDevice.cs]: I guess the idea here is that `GraphicsResource.DisposeAll()` is taking care of calling `Effect.Dispose()`?

Comment (tgjones) [inline MonoGame.Framework/Graphics/GraphicsDevice.cs]: Yes - I stepped through in the debugger, and the `Effect` I was testing with was disposed in `GraphicsResource.DisposeAll()`.

Comment (tomspilman) [inline MonoGame.Framework/Graphics/GraphicsDevice.cs]: Cool...  so the only issue remaining now is that `GraphicsResource.DisposeAll()` will dispose resources owned by a second `GraphicsDevice`.  Same drill as all these other changes... the static collection holding all the `GraphicsResource`s really needs to be owned by the `GraphicsDevice` instance.

Comment (tgjones) [inline MonoGame.Framework/Graphics/GraphicsDevice.cs]: Yep, I have a commit with exactly that fix sitting in a local branch already. I'll push it up as soon as this one is merged - just don't like to have too many PRs open at one time, it becomes difficult to manage.

**Pull request merged**